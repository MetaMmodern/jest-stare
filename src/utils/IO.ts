import * as fs from "fs";
import * as  mkdirp from "mkdirp";
const pkgUp = require("pkg-up");

/**
 * class for simple IO
 * @export
 * @class IO
 */
export class IO {

    /**
     * Delete a file if it exists
     * @static
     * @param {string} file - file to delete
     * @memberof IO
     */
    public static unlinkSync(file: string) {
        if (IO.existsSync(file)) {
            fs.unlinkSync(file);
        }
    }

    /**
     * Wrap fs.writeFile for promise-based calling
     * @static
     * @param {string} path - place to write to
     * @param {*} data - content to write
     * @returns {Promise<void>} - when write is complete
     * @memberof IO
     */
    public static writeFile(path: string, data: any): Promise<void> {
        return new Promise<void>( (resolve, reject) => {
            fs.writeFile(path, data, (error) => {
                if (error) {
                    reject(error);
                }
                resolve();
            });
        });
    }

    /**
     * Wrap fs.writeFile for promise-based calling
     * @static
     * @param {string} path - place to write to
     * @param {*} data - content to write
     * @returns {Promise<void>} - when write is complete
     * @memberof IO
     */
    public static writeFileSync(path: string, data: any) {
        fs.writeFileSync(path, data);
    }

    /**
     * Make directories (recursive)
     * @static
     * @param {string} dir - directory to make
     * @memberof IO
     */
    public static mkdirsSync(dir: string) {
        if (!IO.existsSync(dir)) {
            mkdirp.sync(dir);
        }
    }

    /**
     * Read file and return to caller
     * @static
     * @param {string} path - path to read file
     * @returns - file contents
     * @memberof IO
     */
    public static readFileSync(path: string) {
        return fs.readFileSync(path).toString();
    }

    /**
     * Return whether a given file already exists
     * @static
     * @param {string} path - path of file to check
     * @returns - true if file exists, false otherwise
     * @memberof IO
     */
    public static existsSync(path: string) {
        return fs.existsSync(path);
    }

    /**
     * Read from the user's package.json, if present
     * @static
     * @returns {IJestStareConfig} - config object
     * @returns {object} - full package JSON
     * @memberof IO
     */
    public static readPackageJson(): object {
        const packageJson = pkgUp.sync();
        if (packageJson !== null) {
            return JSON.parse(IO.readFileSync(packageJson));
        } else {
            // if we can't find any package.json, return a blank config
            return {};
        }
    }
}
