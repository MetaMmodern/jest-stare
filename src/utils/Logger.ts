import { format, isNullOrUndefined } from "util";
import * as moment from "moment";

export class Logger {

    public static readonly LEVELS = ["trace", "debug", "info", "warn", "error", "fatal"];
    public static readonly LEVEL_DEFAULT = "debug";

    public static get get() {
        if (isNullOrUndefined(this.mLog)) {
            this.mLog = new Logger();
        }
        return this.mLog;
    }

    public static isValidLevel(level: string) {
        return Logger.LEVELS.indexOf(level) < 0 ? false : true;
    }

    public static validateLevel(level: string) {
        if (!Logger.isValidLevel(level)) {
            throw new Error("invalid level");
        }
    }

    private static mLog: Logger;
    private mPrefix: boolean;
    private mColor: boolean;
    private mIsOn: boolean;
    private mLevel: string;

    constructor() {
        this.mLevel = Logger.LEVEL_DEFAULT;
        this.mPrefix = true;
        this.mColor = true;
        this.mLevel = this.mLevel.toLocaleLowerCase();
        this.mIsOn = true;
        Logger.validateLevel(this.mLevel);
    }

    public isDebugEnabled() {
        return Logger.LEVELS.indexOf("debug") >= Logger.LEVELS.indexOf(this.level) ? true : false;
    }

    public debug(message: string, ...args: any[]) {
        if (!this.isDebugEnabled()) {
            return;
        }
        let adjustedMessage = message;
        if (this.prefix) {
            adjustedMessage = this.buildPrefix("DEBUG") + message;
        }
        if (this.color) {
            // adjustedMessage = TextUtils.chalk.blue(adjustedMessage);
        }
        return this.writeStdout(adjustedMessage, args);
    }

    private writeStderr(message: string, ...args: any[]) {
        const data = this.format(message, args);
        if (this.on) {
            process.stderr.write(this.format(message, args));
        }
        return data;
    }

    private writeStdout(message: string, ...args: any[]) {
        const data = this.format(message, args);
        if (this.on) {
            process.stdout.write(data);
        }
        return data;
    }

    private format(data: string, ...args: any[]) {
        let formatted = data;
        // TODO(Kelosky): this is not ideal, but works for simple cases of
        // .debug(%s, "sub string").
        if (this.formatEnabled && !isNullOrUndefined(args) && args.length > 0) {
            let defined = false;
            args.forEach((arg) => {
                arg.forEach((ntry: string[]) => {
                    if (ntry.length > 0) {
                        defined = true;
                    }
                });
            });
            // if every argument is undefined, dont format it
            if (defined) {
                formatted = format(data, args);
            }
        }
        return formatted + "\n";
    }

    private buildPrefix(type: string) {
        return "[" + moment().format("YYYY/MM/DD HH:MM:SS") + "]" + " " + "[" + type + "]" + " ";
    }

    get formatEnabled() {
        return true;
    }

    set color(isEnabled: boolean) {
        this.mColor = isEnabled;
    }

    get color(): boolean {
        return this.mColor;
    }

    set on(isOn: boolean) {
        this.mIsOn = isOn;
    }

    get on() {
        return this.mIsOn;
    }

    get level() {
        return this.mLevel;
    }

    get prefix(): boolean {
        return this.mPrefix;
    }

    set prefix(prefix) {
        this.mPrefix = prefix;
    }
}
