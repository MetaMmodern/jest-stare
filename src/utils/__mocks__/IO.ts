import * as fs from "fs";

/**
 * class for simple IO
 * @export
 * @class IO
 */
export class IO {

    /**
     * Wrap fs.writeFile for promise-based calling
     * @static
     * @param {string} path - place to write to
     * @param {*} data - content to write
     * @returns {Promise<void>} - when write is complete
     * @memberof IO
     */
    public static writeFile(path: string, data: any): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            resolve();
        });
    }

    /**
     * Make directories (recursive)
     * @static
     * @param {string} dir - directory to make
     * @memberof IO
     */
    public static mkdirsSync(dir: string) {
        return;
    }

    /**
     * Read file and return to caller
     * @static
     * @param {string} path - path to read file
     * @returns - file contents
     * @memberof IO
     */
    public static readFileSync(path: string) {
        throw new Error("Mock read not implemented yet");
    }

    /**
     * Return whether a given file already exists
     * @static
     * @param {string} path - path of file to check
     * @returns - true if file exists, false otherwise
     * @memberof IO
     */
    public static existsSync(path: string) {
        return true;
    }
}
