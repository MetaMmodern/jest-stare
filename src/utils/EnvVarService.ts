import { IJestStareConfig } from "../processor/doc/IJestStareConfig";
import { Logger } from "../utils/Logger";

/**
 * Service for reading environmental variables
 * @export
 * @class EnvVarService
 */
export class EnvVarService {

    /**
     * Creates an instance of EnvVarService.
     * @param {string} mPrefix - prefix of this env var
     * @memberof EnvVarService
     */
    constructor(private mPrefix: string) {}

    /**
     * Read a jest-stare environmental variable's value
     * @private
     * @param {string} suffix - the suffix of the environmental variable to read e.g. JSON_ONLY
     * @returns {string} - env value
     * @memberof EnvVarService
     */
    public readEnvValue(suffix: string): string {
        return process.env[this.mPrefix + suffix];
    }

    /**
     * Determine whether a boolean-type environmental variable was set
     * @private
     * @param {string} envVariableSuffix suffix of the boolean type environmental variable to check
     * @returns {boolean} - whether or not set
     * @memberof EnvVarService
     */
    public readBoolEnvValue(suffix: string): boolean {
        const value = this.readEnvValue(suffix);
        if (value == null) {
            return undefined;
        }
        if ((value).toUpperCase() === "TRUE" || (value === "1")) {
            return true;
        } else if ((value).toUpperCase() === "FALSE" || (value === "0")) {
            return false;
        } else {
            return undefined;
        }
    }
}
