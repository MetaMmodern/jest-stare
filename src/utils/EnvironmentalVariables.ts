import { IJestStareConfig } from "../processor/doc/IJestStareConfig";
import { Logger } from "../utils/Logger";

/**
 * Service for reading environmental variables
 * @export
 * @class EnvironmentalVariables
 */
export class EnvironmentalVariables {

    /**
     * Prefix for all jest-stare related environmental variables
     * @static
     * @type {string}
     * @memberof EnvironmentalVariables
     */
    public static readonly ENV_PREFIX: string = "JEST_STARE_";

    /**
     * Read all jest-stare related environmental variables
     * @returns {IJestStareConfig} - jest stare config from env vars
     * @memberof EnvironmentalVariables
     */
    public read(): IJestStareConfig {
        const additionalResultsProcessorsValue = this.readEnvValue("ADDITIONAL_RESULTS_PROCESSORS");
        let additionalResultsProcessors: string[];
        if (additionalResultsProcessorsValue != null) {
            try {
                additionalResultsProcessors = JSON.parse(additionalResultsProcessorsValue);
            } catch (e) {
                Logger.get.error("Could not parse additional results processors value." +
                    "It should be a JSON string of an array of strings, like the following: " +
                    JSON.stringify(["jest-html-reporter"]) + "\n You specified: " + additionalResultsProcessorsValue);
            }
        }
        return {
            resultDir: this.readEnvValue("RESULT_DIR"),
            resultJson: this.readEnvValue("RESULT_JSON"),
            resultHtml: this.readEnvValue("RESULT_HTML"),
            jsonOnly: this.readBoolEnvValue("JSON_ONLY"),
            jestStareResults: this.readBoolEnvValue("RESULTS"),
            log: this.readBoolEnvValue("LOG"),
            jestStareConfigJson: this.readEnvValue("CONFIG_JSON"),
            coverageLink: this.readEnvValue("COVERAGE_LINK"),
            additionalResultsProcessors
        };
    }

    /**
     * Combine jest stare config from package json and environmental variables,
     * with environmental variables taking precedence
     * @param packageJsonConfig
     * @param {IJestStareConfig} packageJsonConfig - package json config
     * @param {IJestStareConfig} envConfig - env based config
     * @returns {IJestStareConfig} - merged config
     * @memberof EnvironmentalVariables
     */
    public resolve(packageJsonConfig: IJestStareConfig, envConfig: IJestStareConfig): IJestStareConfig {
        const mergedConfig: IJestStareConfig = {};

        if (envConfig.resultDir != null || packageJsonConfig.resultDir != null) {
            mergedConfig.resultDir = envConfig.resultDir == null ? packageJsonConfig.resultDir : envConfig.resultDir;
        }

        if (envConfig.resultJson != null || packageJsonConfig.resultJson != null) {
            mergedConfig.resultJson = envConfig.resultJson == null ? packageJsonConfig.resultJson : envConfig.resultJson;
        }

        if (envConfig.resultHtml != null || packageJsonConfig.resultHtml != null) {
            mergedConfig.resultHtml = envConfig.resultHtml == null ? packageJsonConfig.resultHtml : envConfig.resultHtml;
        }

        if (envConfig.jsonOnly != null || packageJsonConfig.jsonOnly != null) {
            mergedConfig.jsonOnly = envConfig.jsonOnly == null ? packageJsonConfig.jsonOnly : envConfig.jsonOnly;
        }

        if (envConfig.jestStareConfigJson != null || packageJsonConfig.jestStareConfigJson != null) {
            mergedConfig.jestStareConfigJson =
                envConfig.jestStareConfigJson == null ? packageJsonConfig.jestStareConfigJson : envConfig.jestStareConfigJson;
        }

        if (envConfig.log != null || packageJsonConfig.log != null) {
            mergedConfig.log = envConfig.log == null ? packageJsonConfig.log : envConfig.log;
        }

        if (envConfig.jestStareConfigJson != null || packageJsonConfig.jestStareConfigJson != null) {
            mergedConfig.jestStareConfigJson =
                envConfig.jestStareConfigJson == null ? packageJsonConfig.jestStareConfigJson : envConfig.jestStareConfigJson;
        }

        if (envConfig.coverageLink != null || packageJsonConfig.coverageLink != null) {
            mergedConfig.coverageLink = envConfig.coverageLink == null ? packageJsonConfig.coverageLink : envConfig.coverageLink;
        }

        if (envConfig.additionalResultsProcessors != null || packageJsonConfig.additionalResultsProcessors != null) {
            mergedConfig.additionalResultsProcessors =
                envConfig.additionalResultsProcessors == null ? packageJsonConfig.additionalResultsProcessors : envConfig.additionalResultsProcessors;
        }

        return mergedConfig;
    }


    /**
     * Read a jest-stare environmental variable's value
     * @private
     * @param {string} suffix - the suffix of the environmental variable to read e.g. JSON_ONLY
     * @returns {string} - env value
     * @memberof EnvironmentalVariables
     */
    private readEnvValue(suffix: string): string {
        return process.env[EnvironmentalVariables.ENV_PREFIX + suffix];
    }

    /**
     * Determine whether a boolean-type environmental variable was set
     * @private
     * @param {string} envVariableSuffix suffix of the boolean type environmental variable to check
     * @returns {boolean} - whether or not set
     * @memberof EnvironmentalVariables
     */
    private readBoolEnvValue(suffix: string): boolean {
        const value = this.readEnvValue(suffix);
        if (value == null) {
            return undefined;
        }
        if ((value).toUpperCase() === "TRUE" || (value === "1")) {
            return true;
        } else if ((value).toUpperCase() === "FALSE" || (value === "0")) {
            return false;
        } else {
            return undefined;
        }
    }
}
