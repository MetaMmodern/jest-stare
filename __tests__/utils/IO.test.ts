import Mock = jest.Mock;
jest.mock("fs");
import * as fs from "fs";
import { IO } from "../../src/utils/IO";

describe("IO tests", () => {

    beforeEach(() => {
        jest.resetAllMocks();
    });

    it("should pretend to write to a file", async () => {
        const fn = (fs.writeFile as any) as Mock<typeof fs.writeFile>;
        fn.mockImplementation((path: string, data: any, callback: () => void) => {
            callback();
        });

        let error;
        try {
            await IO.writeFile("/place/to/write", "some garbage text");
        } catch (thrownError) {
            error = thrownError;
        }
        expect(error).toBeUndefined();
    });

    it("should match snapshot", () => {
        const data = {
            name: "someone",
            password: "secret",
        };
        expect(data).toMatchSnapshot();
    });

});
