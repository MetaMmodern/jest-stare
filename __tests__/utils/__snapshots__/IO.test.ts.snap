// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IO tests should match snapshot 1`] = `
Object {
  "name": "someone",
  "password": "secret",
}
`;
